`master` build fails with `-std=c++17` Apple gcc flag

I need to use a polymorphic allocator supported in C++17, but NEST build fails with this:

If the sli subsystem expert can take a look at this, that'd be great. Thanks!

```
[ 32%] Building CXX object nestkernel/CMakeFiles/nestkernel.dir/nest_names.cpp.o
[ 32%] Building CXX object nestkernel/CMakeFiles/nestkernel.dir/nestmodule.cpp.o
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1407:52: error: implicit instantiation of undefined template 'std::tuple_size<lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype>>'
_LIBCPP_INLINE_VAR constexpr size_t tuple_size_v = tuple_size<_Tp>::value;
                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: note: in instantiation of variable template specialization 'std::tuple_size_v<lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype>>' requested here
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:1999:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
  auto result = apply( param, positions );
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/__tuple:24:50: note: template is declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS tuple_size;
                                                 ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: error: non-type template argument is not a constant expression
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:1999:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
  auto result = apply( param, positions );
                ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1407:52: error: implicit instantiation of undefined template 'std::tuple_size<sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType>>'
_LIBCPP_INLINE_VAR constexpr size_t tuple_size_v = tuple_size<_Tp>::value;
                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: note: in instantiation of variable template specialization 'std::tuple_size_v<sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType>>' requested here
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:2014:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
  auto result = apply( param, nc );
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/__tuple:24:50: note: template is declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS tuple_size;
                                                 ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: error: non-type template argument is not a constant expression
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:2014:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
  auto result = apply( param, nc );
                ^
4 errors generated.
make[2]: *** [nestkernel/CMakeFiles/nestkernel.dir/nestmodule.cpp.o] Error 1
make[1]: *** [nestkernel/CMakeFiles/nestkernel.dir/all] Error 2
make: *** [all] Error 2
```

--- a/nestkernel/nestmodule.cpp
+++ b/nestkernel/nestmodule.cpp
@@ -1996,7 +1996,7 @@ NestModule::Apply_P_DFunction::execute( SLIInterpreter* i ) const
   auto positions = getValue< DictionaryDatum >( i->OStack.pick( 0 ) );
   auto param = getValue< ParameterDatum >( i->OStack.pick( 1 ) );
 
-  auto result = apply( param, positions );
+  auto result = nest::apply( param, positions ); // Using nest namespace explicitly to avoid confusion with std in C++17
 
   i->OStack.pop( 2 );
   i->OStack.push( result );
@@ -2011,7 +2011,7 @@ NestModule::Apply_P_gFunction::execute( SLIInterpreter* i ) const
   NodeCollectionDatum nc = getValue< NodeCollectionDatum >( i->OStack.pick( 0 ) );
   ParameterDatum param = getValue< ParameterDatum >( i->OStack.pick( 1 ) );
 
-  auto result = apply( param, nc );
+  auto result = nest::apply( param, nc ); // Using nest namespace explicitly to avoid confusion with std in C++17
 
   i->OStack.pop( 2 );
   i->OStack.push( result );

--- a/testsuite/pytests/test_nodeParametrization.py
+++ b/testsuite/pytests/test_nodeParametrization.py
@@ -512,13 +512,13 @@ class TestNodeParametrization(unittest.TestCase):
         """Test cosine of a parameter"""
         for value in np.linspace(-5.0, 5.0, 15):
             p = nest.CreateParameter('constant', {'value': value})
-            self.assertEqual(nest.math.cos(p).GetValue(), np.cos(value))
+            self.assertAlmostEqual(nest.math.cos(p).GetValue(), np.cos(value))
 
     def test_sin_parameter(self):
         """Test sine of a parameter"""
         for value in np.linspace(-5.0, 5.0, 15):
             p = nest.CreateParameter('constant', {'value': value})
-            self.assertEqual(nest.math.sin(p).GetValue(), np.sin(value))
+            self.assertAlmostEqual(nest.math.sin(p).GetValue(), np.sin(value))
 
     def test_power_parameter(self):
         """Test parameter raised to the power of an exponent"""

